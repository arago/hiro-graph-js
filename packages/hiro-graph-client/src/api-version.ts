const APP_API_BASE = '/api/app/7.0';
const AUTH_API_BASE = '/api/auth/6.1';
const EVENTS_WS_API_BASE = '/api/events-ws/6.1';
const GRAPH_API_BASE = '/api/graph/7.1';
const GRAPH_WS_API_BASE = '/api/graph-ws/6.1';
const IAM_API_BASE = '/api/iam/6.1';
const KI_API_BASE = '/api/ki/6';
const VARIABLES_API_BASE = '/api/variables/6';

export {
  APP_API_BASE,
  AUTH_API_BASE,
  EVENTS_WS_API_BASE,
  GRAPH_API_BASE,
  GRAPH_WS_API_BASE,
  IAM_API_BASE,
  KI_API_BASE,
  VARIABLES_API_BASE,
};
