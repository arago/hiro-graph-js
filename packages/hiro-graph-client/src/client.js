/**
 *  A GraphIT Client
 *
 *  It will use Websockets where possible and fall back to HTTP where WebSockets not supported.
 *  Event streams require WebSockets so those will not work witout websocket support.
 *  The connection requires a `Token`.
 */
import WebsocketTransport, { webSocketsAvailable } from './transport-websocket';
import HttpTransport from './transport-http';
import {
    isUnauthorized,
    isTransactionFail,
    isConflict,
    isNotFound,
} from './errors';
import { fixedToken } from './token';
import EventStream from './eventstream';
import authServlet from './servlets/auth';
import apiServlet from './servlets/api';

const passthru = (fn) => [
    (r) => (fn(), r),
    (e) => {
        fn();
        throw e;
    },
];

//nb this isn't a deep clone, just a top level deref.
//we only deal with primitives, arrays and plain objects.
const dereference = (obj) => {
    if (typeof obj !== 'object' || obj === null) {
        return obj;
    }

    if (Array.isArray(obj)) {
        return obj.slice();
    }

    return Object.assign({}, obj);
};

export default class Client {
    constructor({ endpoint, token }, transportOptions = {}, proxies = []) {
        this.endpoint = endpoint;

        //we hold on to the token for ease of access/manual invalidation
        this.setToken(token);

        //create the transports.
        //the http is for http only endpoint, e.g. the /_ki/* and /_variable/* servlets
        this.http = new HttpTransport(endpoint);

        // This is the main transport for most of the api.
        // Use Websocket if available (except if specifically disabled)
        // fallback to http, or use given transport
        // duck type the final parameter
        if (typeof transportOptions.request === 'function') {
            this.transport = transportOptions;
        } else if (webSocketsAvailable && !transportOptions.forceHTTP) {
            //All a transport needs to implement is "request"
            this.transport = new WebsocketTransport(endpoint, transportOptions);
        } else {
            if (!webSocketsAvailable) {
                console.warn(
                    'WebSockets not available, falling back to HTTP transport',
                );
            }

            this.transport = this.http;
        }

        // Bind our fetch for extension servlets.
        this.fetch = (...args) => this.http.fetch(this.token, ...args);
        this.proxyFetch = (proxy) => (url, ...args) =>
            this.fetch(`${proxy}${url}`, ...args);

        this._dedup = Object.create(null);

        //keep this so we can duplicate them
        this._servlets = []; // todo: change _servlets to PlainObject. move custom servlets from this to this._servlets. check getters and setters

        // Auth API
        this.addServlet(
            'auth',
            authServlet,
            proxies.length >= 1 ? proxies[0] : '',
        );

        // Global API
        this.addServlet(
            'api',
            apiServlet,
            proxies.length >= 2 ? proxies[1] : '',
        );
    }

    setToken(token) {
        this.token = typeof token === 'string' ? fixedToken(token) : token;
    }

    // NB this is not held anywhere in this instance, but returned
    // to the caller. It only connects when it's subscribe() method
    // is called.
    eventStream(filters = [], { groupId, offset, scopeId } = {}) {
        let filtersArray = filters;

        if (!Array.isArray(filters)) {
            filtersArray = [filters];
        }

        if (filtersArray.some((f) => typeof f !== 'string')) {
            throw new Error(
                'All filters must be strings representing `jfilter` filters',
            );
        }

        if (filtersArray.length === 0) {
            // add a default one that catches everything
            filtersArray.push('(element.ogit/_id = *)');
        }

        return new EventStream(
            { endpoint: this.endpoint, token: this.token },
            { groupId, offset, scopeId, filters: filtersArray },
        );
    }

    /**
     *  Make a clone of this connection, but with new
     */
    cloneWithNewToken(newToken) {
        const cloned = new Client(
            {
                endpoint: this.endpoint,
                token: newToken,
            },
            this.transport,
        );

        this._servlets.forEach(([name, methods]) =>
            cloned.addServlet(name, methods),
        );

        return cloned;
    }

    /**
     *  Make a generic request. The transport will handle the details, we handle the retry.
     *  We use `this.transport` which is websocket if available.
     */
    request(
        { type, headers = {}, body = {} } = {},
        reqOptions = {},
        retries = 1,
    ) {
        return this.transport.request(
            this.token,
            { type, headers, body },
            reqOptions,
        );
    }

    // Deduplicates a request, but the request must be an Array returning one.
    dedupedRequest(
        { type, headers = {}, body = {} } = {},
        reqOptions = {},
        retries = 1,
    ) {
        try {
            const requestKey = JSON.stringify({ type, headers, body });

            if (requestKey in this._dedup) {
                this._dedup[requestKey]._calls++;

                return this._dedup[requestKey].then(dereference);
            }

            const cleanUp = passthru(() => {
                const calls = this._dedup[requestKey]._calls;

                delete this._dedup[requestKey];
            });
            const promise = (this._dedup[requestKey] = this.request(
                { type, headers, body },
                reqOptions,
                retries,
            )
                .then(...cleanUp)
                .then((res) => (promise._calls > 1 ? dereference(res) : res)));

            promise._calls = 1;

            return promise;
        } catch (e) {
            return Promise.reject(e);
        }
    }

    getToken() {
        return this.token;
    }

    /**
     *  Core GraphIT REST API methods
     *  Prefer websocket transport
     *  The servlets for e.g. /_ki/validate are added using mixins
     */

    /**
     *  Get a single item by ID
     */
    get(id, reqOptions = {}) {
        const headers = { 'ogit/_id': id };

        if (reqOptions.listMeta) {
            headers.listMeta = true;
        }

        return this.dedupedRequest(
            { type: 'get', headers },
            reqOptions,
        );
    }

    /**
     *  Get the node for the owner of this token
     */
    me(reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: 'getme',
                body: {},
                headers: { profile: true, 'me-type': 'account' },
            },
            reqOptions,
        );
    }

    /**
     *  Create a new node
     */
    create(type, data = {}, reqOptions = {}) {
        const headers = { 'ogit/_type': type };

        if (reqOptions.waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return this.request(
            { type: 'create', headers, body: data },
            reqOptions,
        );
    }

    /**
     *  Update a Node
     */
    update(id, data = {}, reqOptions = {}) {
        const headers = { 'ogit/_id': id };

        if (reqOptions.waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return this.request(
            { type: 'update', headers, body: data },
            reqOptions,
        );
    }

    /**
     *  Replace a Node, optionally `upsert`
     */
    replace(id, data = {}, reqOptions = {}) {
        const { createIfNotExists = false, waitForIndex = false } = reqOptions;
        //createIfNotExists should contain the "ogit/_type" to create if the node doesn't exist,
        //and nothing otherwise
        const headers = { 'ogit/_id': id };

        if (createIfNotExists) {
            Object.assign(headers, {
                createIfNotExists: 'true',
                'ogit/_type': createIfNotExists,
            });
        }

        if (waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return his.request(
            { type: 'replace', headers, body: data },
            reqOptions,
        );
    }

    /**
     *  Delete a node/edge
     *
     *  We don't handle the 404/409 here. or should we?
     */
    delete(id, reqOptions = {}) {
        const { waitForIndex = false } = reqOptions;
        const headers = { 'ogit/_id': id };

        if (waitForIndex) {
            headers.waitForIndex = 'true';
        }

        return this.request({ type: 'delete', headers }, reqOptions);
    }

    /**
     *  This is a vertices query
     */
    lucene(
        query = '',
        {
            limit = 50,
            offset = 0,
            order = false,
            fields = [],
            count = false,
            ...placeholders
        } = {},
        reqOptions = {},
    ) {
        const body = {
            query,
            limit,
            offset,
            ...placeholders,
        };

        if (count) {
            body.count = 'true'; // string true
        }

        if (order) {
            body.order = '' + order; // this implicitly does array.join(",") for arrays. (and works with strings...)
        }

        if (fields.length > 0) {
            body.fields = Array.isArray(fields)
                ? fields.join(',')
                : String(fields);
        }

        return this.dedupedRequest(
            {
                type: 'query',
                headers: { type: 'vertices' },
                body,
            },
            reqOptions,
        );
    }

    ids(list, reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: 'query',
                headers: { type: 'ids' },
                body: { query: list.join(',') }, // yes, it has to be a comma-seperated string
            },
            reqOptions,
        );
    }

    /**
     *  This is a gremlin query
     */
    gremlin(root, query, reqOptions = {}) {
        return this.dedupedRequest(
            {
                type: 'query',
                headers: { type: 'gremlin' },
                body: { root, query },
            },
            reqOptions,
        );
    }

    /**
     *  Connect two Nodes with an edge of `type`
     */
    connect(type, inId, outId, reqOptions = {}) {
        returnthis
            .request(
                {
                    type: 'connect',
                    headers: { 'ogit/_type': type },
                    body: { in: inId, out: outId },
                },
                reqOptions,
            )
            .then(
                () => {}, //return nothing.
                (err) => {
                    //Conflict is OK here, just means that the edge was already connected.
                    if (isConflict(err)) {
                        return; //return nothing.
                    }

                    //real error.
                    throw err;
                },
            );
    }

    /**
     *  Disconnect two nodes, convenience for delete, generates the edge id for you.
     */
    disconnect(type, inId, outId, reqOptions = {}) {
        return this.delete(`${outId}$$${type}$$${inId}`, reqOptions).then(
            () => {}, //return nothing.
            (err) => {
                //Not Found or Conflict is OK here, just means that the edge was already deleted/didn't ever exist
                if (isNotFound(err) || isConflict(err)) {
                    return; //return nothing.
                }

                //real error.
                throw err;
            },
        );
    }

    /**
     *  Write timeseries values (only ogit/Timeseries vertices)
     *
     *  values are { timestamp: millisecond unix, value: string value }
     */
    writets(timeseriesId, values) {
        let items = values;

        if (!Array.isArray(values)) {
            items = [values];
        }

        return this.request({
            type: 'writets',
            headers: {
                'ogit/_id': timeseriesId,
            },
            body: { items },
        });
    }

    /**
     *  Read timeseries values (only ogit/Timeseries vertices)
     */
    streamts(
        timeseriesId,
        opts = {
            from: false,
            to: false,
            limit: 50,
            includeDeleted: false,
            with: [],
        },
    ) {
        const headers = {
            'ogit/_id': timeseriesId,
            ...opts,
        };
        const body = {};

        return this.dedupedRequest({
            type: 'streamts',
            headers,
            body,
        });
    }

    /**
     * Placeholders here for reading/writing log files (ogit/Log)
     */
    //eslint-disable-next-line no-unused-vars
    writelog(logId, entries) {
        return Promise.reject(new Error('Log writing is not implemented yet'));
    }
    //eslint-disable-next-line no-unused-vars
    readlog(logId, options) {
        return Promise.reject(new Error('Log reading is not implemented yet'));
    }

    /**
     *  Returns previous versions of a vertex
     */
    history(
        id,
        {
            offset = false,
            limit = false,
            from = false,
            to = false,
            version = false,
            type = false,
            listMeta = false,
            includeDeleted = false,
            vid = false,
        } = {},
    ) {
        const headers = { 'ogit/_id': id, listMeta, includeDeleted };
        const body = {};

        if (offset !== false) {
            headers.offset = offset;
        }

        if (limit !== false) {
            headers.limit = limit;
        }

        if (from !== false) {
            headers.from = from;
        }

        if (to !== false) {
            headers.to = to;
        }

        if (version !== false) {
            headers.version = version;
        }

        if (type !== false) {
            headers.type = type;
        }

        if (vid !== false) {
            headers.vid = vid;
        }

        return this.dedupedRequest({
            type: 'history',
            headers: headers,
            body,
        });
    }

    /**
     *  Custom servlet endpoints
     *  always use HTTP transport, which knows how to handle these types.
     *
     *  The servletMethods are passed:
     *  - `fetch` a token/endpoint aware HTTP fetch API.
     *  - `options` the default fetch options you can override
     *  - `...args` the rest of the args passed in by the user for that method
     */
    addServlet(prefix, servletMethods, proxy) {
        if (!prefix) {
            throw new Error('[GRAPH] Must give prefix for servlet');
        }

        if (prefix in this) {
            throw new Error(
                '[GRAPH] Sevlet Extensions must have unique prefixes. Attempted to re-add `' +
                    prefix +
                    '`',
            );
        }

        const fetch = proxy ? this.proxyFetch(proxy) : this.fetch;
        const isFactory = typeof servletMethods === 'function';

        const servletDefinition = isFactory
            ? servletMethods(fetch, this.http.defaultOptions())
            : servletMethods;

        //create namespace.
        this[prefix] = {};
        //add servlet methods
        Object.keys(servletDefinition).reduce((acc, method) => {
            acc[method] = (...args) => {
                const servletMethod = servletDefinition[method];

                const legacyArgs = isFactory
                    ? []
                    : [fetch, this.http.defaultOptions()];

                const callArgs = [...legacyArgs, ...args];

                return servletMethod(...callArgs).catch((err) => {
                    //these are the special cases.
                    //regular errors end up with code === undefined, so not retryable.
                    switch (true) {
                        case isUnauthorized(err): //unauthorized (which means unauthenticated) invalidate TOKEN.
                            this.token.invalidate();
                            err.isRetryable = true;
                            break;
                        case isTransactionFail(err): //error persisting transaction. retryable.
                            err.isRetryable = true;
                            break;
                        //there are other known errors, e.g. 403, 400, etc... but they are not retryable.
                        default:
                            if (err.isRetryable === undefined) {
                                err.isRetryable = false;
                            }

                            break;
                    }

                    //a chance to retry - only once.
                    if (err.isRetryable) {
                        return servletMethod(...callArgs);
                    }

                    throw err;
                });
            };

            return acc;
        }, this[prefix]);

        //chainable? might be useful.
        return this;
    }

    getServlet(prefix) {
        return this[prefix];
    }
}
