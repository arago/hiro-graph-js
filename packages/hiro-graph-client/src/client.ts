/**
 *  A GraphIT Client
 *
 *  It will use Websockets where possible and fall back to HTTP where WebSockets not supported.
 *  Event streams require WebSockets so those will not work witout websocket support.
 *  The connection requires a `Token`.
 */
import { catchError, map, retryWhen, concatMap, delay } from 'rxjs/operators';
import { of, throwError, EMPTY } from 'rxjs';

import {
  WebSocketTransport,
  webSocketsAvailable,
  WebSocketRequestOptions,
} from './transport-websocket';
import { HttpTransport } from './transport-http';
import * as Errors from './errors';
import { fixedToken } from './token';
import { EventStream, EventStreamRequest } from './eventstream';
import {
  GraphTransport,
  GraphRequest,
  TimeseriesResponse,
  HistoryQueryOptions,
} from './types';
import {
  GremlinQueryFunction,
  GremlinQuery,
  gremlin as CreateGremlin,
} from './gremlin';
import { Lucene, LuceneQueryOptions, lucene as CreateLucene } from './lucene';

import { Token } from '.';

export interface ClientOptions {
  endpoint: string;
  token: string | Token;
}

export type TransportOrOptions = WebSocketRequestOptions | GraphTransport;

const isTransport = (t: TransportOrOptions): t is GraphTransport =>
  'request' in t && typeof t.request === 'function';

export class Client {
  private endpoint: string;
  private http: HttpTransport;
  private transport: GraphTransport;
  private token: Token;

  constructor(
    { endpoint, token }: ClientOptions,
    transportOrOptions: TransportOrOptions = {},
  ) {
    this.endpoint = endpoint;

    this.token = this.setToken(token);

    //create the transports.
    //the http is for http only endpoint, e.g. the /_ki/* and /_variable/* servlets
    this.http = new HttpTransport(endpoint);

    // This is the main transport for most of the api.
    // Use Websocket if available (except if specifically disabled)
    // fallback to http, or use given transport
    // duck type the final parameter
    if (isTransport(transportOrOptions)) {
      this.transport = transportOrOptions;
    } else if (webSocketsAvailable && !transportOrOptions.forceHTTP) {
      //All a transport needs to implement is "request"
      this.transport = new WebSocketTransport(endpoint, transportOrOptions);
    } else {
      if (!webSocketsAvailable) {
        console.warn(
          'WebSockets not available, falling back to HTTP transport',
        );
      }

      this.transport = this.http;
    }
  }

  setToken(token: string | Token) {
    this.token = typeof token === 'string' ? fixedToken(token) : token;

    return this.token;
  }

  // NB this is not held anywhere in this instance, but returned
  // to the caller. It only connects when it's subscribe() method
  // is called.
  eventStream(
    filters: string | string[] = [],
    { groupId, offset, scopeId }: Omit<EventStreamRequest, 'filters'> = {},
  ) {
    const filtersArray = Array.isArray(filters) ? filters : [filters];

    if (filtersArray.length === 0) {
      // add a default one that catches everything
      filtersArray.push('(element.ogit/_id = *)');
    }

    return new EventStream(
      { endpoint: this.endpoint, token: this.token },
      { groupId, offset, scopeId, filters: filtersArray },
    );
  }

  /**
   *  Make a clone of this connection, but with new
   */
  cloneWithNewToken(newToken: Token) {
    const cloned = new Client(
      {
        endpoint: this.endpoint,
        token: newToken,
      },
      this.transport,
    );

    return cloned;
  }

  /**
   *  Make a generic request. The transport will handle the details, we handle the retry.
   *  We use `this.transport` which is websocket if available.
   */
  request<T>({ type, headers = {}, body = {} }: GraphRequest, maxRetries = 1) {
    let retries = 0;

    return this.transport
      .request<T>(this.token, { type, headers, body })
      .pipe(
        retryWhen((err$) =>
          err$.pipe(
            concatMap((res) => {
              const closed = res === Errors.connectionClosedBeforeSend;

              // Always handle early close
              if (closed) {
                return of(res);
              }

              if (res.code === 401) {
                this.token.invalidate();
              }

              // Retry retryable error until max retries
              if (
                retries < maxRetries &&
                (res.isRetryable || res.code === 888 || res.code === 401)
              ) {
                retries += 1;

                return of(res);
              }

              return throwError(res);
            }),
            delay(1000),
          ),
        ),
      );
  }

  getToken() {
    return this.token;
  }

  /**
   *  Core GraphIT REST API methods
   *  Prefer websocket transport
   *  The servlets for e.g. /_ki/validate are added using mixins
   */

  /**
   *  Get a single item by ID
   */
  get<T>(id: string, options: { listMeta?: boolean } = {}) {
    const headers = { 'ogit/_id': id } as any;

    if (options.listMeta) {
      headers.listMeta = true;
    }

    return this.request<T>({ type: 'get', headers });
  }

  /**
   *  Get the node for the owner of this token
   */
  me<T>() {
    return this.request<T>({
      type: 'getme',
      body: {},
      headers: { profile: true, 'me-type': 'account' },
    });
  }

  /**
   *  Create a new node
   */
  create<T>(type: string, data = {}, options: { waitForIndex?: boolean } = {}) {
    const headers: Record<string, string> = { 'ogit/_type': type };

    if (options.waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'create', headers, body: data });
  }

  /**
   *  Update a Node
   */
  update<T>(id: string, data = {}, options: { waitForIndex?: boolean } = {}) {
    const headers: Record<string, string> = { 'ogit/_id': id };

    if (options.waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'update', headers, body: data });
  }

  /**
   *  Replace a Node, optionally `upsert`
   */
  replace<T>(
    id: string,
    data: Partial<T> = {},
    options: { waitForIndex?: boolean; createIfNotExists?: string } = {},
  ) {
    const { createIfNotExists = false, waitForIndex = false } = options;
    //createIfNotExists should contain the "ogit/_type" to create if the node doesn't exist,
    //and nothing otherwise
    const headers: Record<string, string> = { 'ogit/_id': id };

    if (createIfNotExists) {
      Object.assign(headers, {
        createIfNotExists: 'true',
        'ogit/_type': createIfNotExists,
      });
    }

    if (waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'replace', headers, body: data });
  }

  /**
   *  Delete a node/edge
   *
   *  We don't handle the 404/409 here. or should we?
   */
  delete<T>(id: string, options: { waitForIndex?: boolean } = {}) {
    const { waitForIndex = false } = options;
    const headers: Record<string, string> = { 'ogit/_id': id };

    if (waitForIndex) {
      headers.waitForIndex = 'true';
    }

    return this.request<T>({ type: 'delete', headers });
  }

  /**
   *  This is a vertices query
   */

  lucene<T>(
    query: string | Lucene.Query = '',
    {
      limit = 50,
      offset = 0,
      order,
      fields = [],
      count = false,
      ...placeholders
    }: LuceneQueryOptions = {},
  ) {
    let _query;
    let _placeholders = placeholders;

    if (typeof query === 'object') {
      const res = CreateLucene(query);

      _query = res.querystring;
      _placeholders = {
        ...placeholders,
        ...res.placeholders,
      };
    } else {
      _query = query;
    }

    const body: {
      query: string;
      limit?: number;
      offset?: number;
      order?: string;
      fields?: string;
      count?: string;
      [index: string]: string | number | undefined;
    } = {
      query: _query,
      limit,
      offset,
      ..._placeholders,
    };

    if (count) {
      body.count = 'true'; // string true
    }

    if (order) {
      body.order = '' + order; // this implicitly does array.join(",") for arrays. (and works with strings...)
    }

    if (fields.length > 0) {
      body.fields = Array.isArray(fields) ? fields.join(',') : String(fields);
    }

    return this.request<T>({
      type: 'query',
      headers: { type: 'vertices' },
      body,
    });
  }

  ids<T>(list: string[]) {
    return this.request<T>({
      type: 'query',
      headers: { type: 'ids' },
      body: { query: list.join(',') }, // yes, it has to be a comma-seperated string
    });
  }

  /**
   *  This is a gremlin query
   */

  gremlin<T>(
    root: string,
    queryOrBuilder: GremlinQuery | GremlinQueryFunction,
  ) {
    const query =
      typeof queryOrBuilder === 'function'
        ? queryOrBuilder(CreateGremlin(''))
        : queryOrBuilder;

    return this.request<T>({
      type: 'query',
      headers: { type: 'gremlin' },
      body: { root, query: query.toString() },
    });
  }

  /**
   *  Connect two Nodes with an edge of `type`
   */
  connect<T>(type: string, inId: string, outId: string) {
    return this.request<T>({
      type: 'connect',
      headers: { 'ogit/_type': type },
      body: { in: inId, out: outId },
    }).pipe(
      catchError((err) => {
        if (Errors.isConflict(err)) {
          return EMPTY;
        }

        throw err;
      }),
      map(() => {}),
    );
  }

  /**
   *  Disconnect two nodes, convenience for delete, generates the edge id for you.
   */
  disconnect<T>(type: string, inId: string, outId: string) {
    return this.delete<T>(`${outId}$$${type}$$${inId}`).pipe(
      catchError((err) => {
        if (Errors.isNotFound(err) || Errors.isConflict(err)) {
          return EMPTY;
        }

        throw err;
      }),
      map(() => {}),
    );
  }

  /**
   *  Write timeseries values (only ogit/Timeseries vertices)
   *
   *  values are { timestamp: millisecond unix, value: string value }
   */
  writets<T>(
    timeseriesId: string,
    values: TimeseriesResponse | TimeseriesResponse[],
  ) {
    const items = Array.isArray(values) ? values : [values];

    return this.request<T>({
      type: 'writets',
      headers: {
        'ogit/_id': timeseriesId,
      },
      body: { items },
    });
  }

  /**
   *  Read timeseries values (only ogit/Timeseries vertices)
   */
  streamts<T>(
    timeseriesId: string,
    options: {
      from?: number | false;
      to?: number | false;
      limit?: number | false;
      includeDeleted?: boolean;
      with?: string[];
    } = {
      from: false,
      to: false,
      limit: 50,
      includeDeleted: false,
      with: [],
    },
  ) {
    const headers = {
      'ogit/_id': timeseriesId,
      ...options,
    };
    const body = {};

    return this.request<T>({
      type: 'streamts',
      headers,
      body,
    });
  }

  /**
   *  Returns previous versions of a vertex
   */
  history<T>(
    id: string,
    {
      listMeta = false,
      includeDeleted = false,
      ...options
    }: HistoryQueryOptions = {},
  ) {
    const headers: Record<string, any> = {
      'ogit/_id': id,
      listMeta,
      includeDeleted,
      ...options,
    };

    return this.request<T>({
      type: 'history',
      headers: headers,
      body: {},
    });
  }
}
