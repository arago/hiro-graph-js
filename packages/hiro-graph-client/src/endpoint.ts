import qs, { ParsedUrlQueryInput } from 'querystring';

export const APIs = {
  http: {
    app: '/api/app/7.0',
    auth: '/api/auth/6.1',
    graph: '/api/graph/7.1',
    iam: '/api/iam/6.1',
    ki: '/api/ki/6',
    variables: '/api/variables/6',
  },
  ws: {
    events: '/api/events-ws/6.1',
    graph: '/api/graph-ws/6.1',
  },
};

export type WS_API = keyof typeof APIs['ws'];
export type HTTP_API = keyof typeof APIs['http'];

export class Endpoint<WS extends boolean = false> {
  private _value: string;
  protected readonly APIs?: Record<string, string>;

  constructor(value: string, isWebsocket?: WS) {
    this._value = value.replace(/\/$/, '');

    if (isWebsocket) {
      this._value = this._value.replace(/^http/, 'ws');
      this.APIs = APIs.ws;
    } else {
      this.APIs = APIs.http;
    }
  }

  get value() {
    return this._value;
  }

  api(
    name: WS extends true ? WS_API : HTTP_API,
    path?: string,
    query?: ParsedUrlQueryInput,
  ) {
    if (!this.APIs) {
      throw new Error(`No APIs defined for this endpoint: ${this.value}`);
    }

    let url = this._value + this.APIs[name];

    if (path) {
      if (path.indexOf('http') === 0) {
        url = path;
      } else {
        let p = path.replace(/\/$/, '');

        if (!p.startsWith('/')) {
          p = `/${p}`;
        }

        url += p;
      }
    }

    if (query && Object.keys(query).length > 0) {
      url += `/?${qs.stringify(query)}`;
    }

    return url;
  }
}
