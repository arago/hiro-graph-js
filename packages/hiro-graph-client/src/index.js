import Token, { cannotGetToken } from './token';
import Client from './client';
import * as Errors from './errors';

export default Client;
export { Token, cannotGetToken, Errors };
