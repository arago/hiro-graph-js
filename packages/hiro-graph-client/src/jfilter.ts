import memoize from 'lodash/memoize';
import omit from 'lodash/omit';

export interface JFilterType {
  toString: () => string;
  test: (obj: any) => boolean;
}

class BaseFilter {
  protected readonly filters: JFilterType[] = [];

  constructor(filters: JFilterType[]) {
    this.filters = filters;
  }
}

class JFilterOR extends BaseFilter implements JFilterType {
  toString() {
    return `|${this.filters.map(wrapParens).join('')}`;
  }

  test(obj: any) {
    return this.filters.some((f: JFilterType) => f.test(obj));
  }
}

class JFilterAND extends BaseFilter implements JFilterType {
  toString() {
    return `&${this.filters.map(wrapParens).join('')}`;
  }

  test(obj: any) {
    return this.filters.every((f: JFilterType) => f.test(obj));
  }
}

class JFilterEQ extends BaseFilter implements JFilterType {
  private readonly prop: any;
  private readonly value: any;

  constructor(prop: any, value: any) {
    super([]);
    this.prop = prop;
    this.value = value;
  }

  toString() {
    return `(${this.prop}=${this.value})`;
  }

  test(obj: any) {
    return matches(obj, this.prop, this.value);
  }
}

class JFilter {
  static or(...filters: any): JFilterType {
    return new JFilterOR(filters);
  }

  static and(...filters: any): JFilterType {
    return new JFilterAND(filters);
  }

  static equals(prop: any, value: any): JFilterType {
    return new JFilterEQ(prop, value);
  }

  static transform(event: any) {
    return omit(
      {
        ...event,
        element: event.body,
        action: event.type,
      },
      ['body', 'type'],
    );
  }
}

const getMemoizedMatcher = memoize((value) => {
  if (!value.includes('*') && !value.includes('?')) {
    return (actual: any) => actual === value;
  }

  const processed = value
    .split(/([?*])/)
    .map((chunk: any, i: any) => {
      if (i % 2) {
        return `.${chunk === '?' ? '{1}' : '{1,}'}`;
      }

      return escapeRegexOperators(chunk);
    })
    .join('');

  const regex = new RegExp(`^${processed}$`);

  return (actual: any) => regex.test(actual);
});

function escapeRegexOperators(str: any) {
  return str.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
}

function getProp(obj: any, propString: any) {
  const keys = propString.split('.');

  return keys.reduce((o: any, k: any) => {
    if (typeof o === 'object' && k in o) {
      return o[k];
    }

    return false;
  }, obj);
}

function wrapParens(filter: any) {
  const inner = filter.toString();

  return inner[0] === '(' ? inner : `(${inner})`;
}

function matches(obj: any, prop: any, value: any) {
  const actual = getProp(obj, prop);
  const matcher = getMemoizedMatcher(value);

  return matcher(actual);
}

/*
  Exports
 */
export { JFilter };
