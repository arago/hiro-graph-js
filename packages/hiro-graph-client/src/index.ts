import { Token, cannotGetToken } from './token';
import { Client, ServletFactory } from './client';
import * as Errors from './errors';

export default Client;
export * from './gremlin';
export * from './lucene';
export * from './types';
export * from './jfilter';
export * from './endpoint';
export * as Servlets from './servlets/index';
export { cannotGetToken, Errors, ServletFactory, Token };
