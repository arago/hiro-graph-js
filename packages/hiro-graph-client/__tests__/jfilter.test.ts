import { JFilter } from '../src/jfilter';

test('JFilter snapshots', () => {
  const tests = [
    JFilter.and(JFilter.equals('a', 'b'), JFilter.equals('d', 'e')),
    JFilter.or(JFilter.equals('a', 'b'), JFilter.equals('d', 'e')),
    JFilter.and(JFilter.equals('a', 'b'), JFilter.equals('d', 'e')),
    JFilter.and(
      JFilter.equals('a', 'b'),
      JFilter.or('e=f'),
      JFilter.and(JFilter.equals('d', 'c'), JFilter.equals('g', 'f')),
    ),
  ];

  tests.map((t) => expect(t.toString()).toMatchSnapshot());
});

test('JFilter functions', () => {
  expect(JFilter.equals('a', 'b').toString()).toBe('(a=b)');
  expect(JFilter.and('a', 'b').toString()).toBe('&(a)(b)');
  expect(JFilter.or('a', 'b').toString()).toBe('|(a)(b)');
});

test('JFilter nested', () => {
  expect(JFilter.and('a', JFilter.and('b', 'c')).toString()).toBe(
    '&(a)(&(b)(c))',
  );
  expect(JFilter.or('a', JFilter.or('b', 'c')).toString()).toBe(
    '|(a)(|(b)(c))',
  );
});

test('JFilter transform', () => {
  expect(
    JFilter.transform({ body: { 'ogit/_id': 'test' }, type: 'UPDATE' }),
  ).toEqual({
    element: { 'ogit/_id': 'test' },
    action: 'UPDATE',
  });
});

test('JFilter test function', () => {
  expect(
    JFilter.equals('element.ogit/_id', 'test').test({
      element: { 'ogit/_id': 'test' },
      action: 'UPDATE',
    }),
  ).toEqual(true);

  expect(
    JFilter.equals('action', 'UPDATE').test({
      element: { 'ogit/_id': 'test' },
      action: 'CREATE',
    }),
  ).toEqual(false);

  expect(
    JFilter.equals('action', '*').test({
      element: { 'ogit/_id': 'test' },
      action: 'CREATE',
    }),
  ).toEqual(true);

  expect(
    JFilter.and(
      JFilter.equals('element.ogit/_id', 'test'),
      JFilter.equals('action', 'CREATE'),
    ).test({
      element: { 'ogit/_id': 'test' },
      action: 'CREATE',
    }),
  ).toEqual(true);

  expect(
    JFilter.and(
      JFilter.equals('element.ogit/_id', 'test'),
      JFilter.equals('action', 'CREATE'),
    ).test({
      element: { 'ogit/_id': 'test2' },
      action: 'CREATE',
    }),
  ).toEqual(false);
});
