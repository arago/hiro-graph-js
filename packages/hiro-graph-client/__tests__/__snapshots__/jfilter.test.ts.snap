// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JFilter snapshots 1`] = `"&(a = b)(d = e)"`;

exports[`JFilter snapshots 2`] = `"|(a = b)(d = e)"`;

exports[`JFilter snapshots 3`] = `"&(a = b)(d = e)"`;

exports[`JFilter snapshots 4`] = `"&(a = b)(|(e = f)&(d = c)(g = f))"`;
